File: julia/scripts/benchmark.jl
using Printf
include("../src/PlutoSim.jl")
using .PlutoSim

function compute_total_energy(system::System)
    KE = 0.5 * sum(system.m[i] * sum(system.v[i, :].^2) for i in 1:system.num_particles)
    PE = 0.0
    for i in 1:system.num_particles
        for j in (i+1):system.num_particles
            r = sqrt(sum((system.x[j, :] .- system.x[i, :]).^2))
            PE -= system.G * system.m[i] * system.m[j] / r
        end
    end
    return KE + PE
end

function bench_integrator(name::String, integrator_fn!, force_evals::Int,
                           ic::String, num_steps::Int, dt::Float64,
                           warmup_steps::Int=10; quiet::Bool=false)
    # Warmup: fresh system, run warmup_steps to trigger JIT
    system, _, _ = PlutoSim.get_initial_conditions(ic)
    a = zeros(system.num_particles, 3)
    PlutoSim.acceleration!(a, system)
    if !quiet
        print("  [$name] warming up ($warmup_steps steps)... ")
        flush(stdout)
    end
    for _ in 1:warmup_steps
        integrator_fn!(system, a, dt)
    end
    if !quiet
        println("done")
    end

    # Fresh system for timed run
    system, _, _ = PlutoSim.get_initial_conditions(ic)
    a = zeros(system.num_particles, 3)
    PlutoSim.acceleration!(a, system)
    E0 = compute_total_energy(system)

    t0 = time()
    for _ in 1:num_steps
        integrator_fn!(system, a, dt)
    end
    wall = time() - t0

    Ef = compute_total_energy(system)
    energy_error = abs((Ef - E0) / E0)
    sps = num_steps / wall
    mean_ms = wall / num_steps * 1000

    if !quiet
        @printf("  [%s] done: %.1f steps/s, |dE/E0|=%.3e, mean=%.4f ms/step\n",
                name, sps, energy_error, mean_ms)
    end

    return Dict(
        "name"                 => name,
        "steps_per_sec"        => sps,
        "mean_ms_per_step"     => mean_ms,
        "energy_error"         => energy_error,
        "force_evals_per_step" => force_evals,
        "wall_s"               => wall,
        "E0"                   => E0,
        "Ef"                   => Ef,
    )
end

function run_benchmarks(; json_mode::Bool=false)
    IC        = "solar_system_plus"
    NUM_STEPS = 1000
    DT        = 5.0

    if !json_mode
        println()
        println("=" ^ 60)
        println("JULIA N-BODY BENCHMARK")
        @printf("Initial condition : %s  (%d steps, dt=%.1f days)\n", IC, NUM_STEPS, DT)
        println("=" ^ 60)
    end

    integrators = [
        ("Velocity-Verlet", PlutoSim.velocity_verlet!, 2),
        ("Ruth-Forest",     PlutoSim.ruth_forest!,     3),
        ("Yoshida-4",       PlutoSim.yoshida4!,        4),
    ]

    results = []
    for (name, fn!, fevals) in integrators
        r = bench_integrator(name, fn!, fevals, IC, NUM_STEPS, DT; quiet=json_mode)
        push!(results, r)
    end

    if json_mode
        # Emit compact JSON to stdout for benchmark_all.py to parse
        entries = join([
            string("{\"integrator\":\"", r["name"],
                   "\",\"steps_per_sec\":", r["steps_per_sec"],
                   ",\"energy_error\":", r["energy_error"], "}")
            for r in results
        ], ",")
        println("{\"results\":[", entries, "]}")
    else
        # Human-readable table
        println()
        println("=" ^ 70)
        @printf("%-20s  %12s  %14s  %18s\n",
                "Integrator", "Steps/sec", "|dE/E0|", "Force evals/step")
        println("-" ^ 70)
        for r in results
            @printf("%-20s  %12.1f  %14.3e  %18d\n",
                    r["name"], r["steps_per_sec"], r["energy_error"],
                    r["force_evals_per_step"])
        end
        println("=" ^ 70)

        # Memory estimate
        system, _, _ = PlutoSim.get_initial_conditions(IC)
        a = zeros(system.num_particles, 3)
        mem_bytes = sizeof(system.x) + sizeof(system.v) + sizeof(system.m) + sizeof(a)
        @printf("\nMemory (state arrays): %d bytes (%.1f KB), %.0f bytes/particle\n\n",
                mem_bytes, mem_bytes/1024, mem_bytes/system.num_particles)
    end

    return results
end

# Detect --json flag
json_mode = "--json" in ARGS
run_benchmarks(; json_mode=json_mode)

File: cpp/src/benchmark.cpp
#include <chrono>
#include <cmath>
#include <cstdio>
#include <functional>
#include <string>
#include <vector>

#include "initial_conditions.hpp"
#include "integrators.hpp"

static double total_energy(const System& sys) {
    double KE = 0.0;
    for (int i = 0; i < sys.N; ++i)
        KE += 0.5 * sys.m[i] * sys.v[i].norm2();

    double PE = 0.0;
    for (int i = 0; i < sys.N; ++i)
        for (int j = i + 1; j < sys.N; ++j) {
            double r = (sys.x[j] - sys.x[i]).norm();
            PE -= sys.G * sys.m[i] * sys.m[j] / r;
        }
    return KE + PE;
}

struct BenchResult {
    std::string name;
    double steps_per_sec;
    double energy_error;
    double wall_time_s;
    int force_evals;
};

static BenchResult run_bench(
    const std::string& name,
    int force_evals,
    int steps,
    int warmup_steps,
    double dt,
    std::function<void(System&, std::vector<Vec3>&, double)> step_fn)
{
    // Warmup pass (ensures branch predictors / caches are warm)
    {
        System sys = make_solar_system();
        std::vector<Vec3> a = compute_acceleration(sys);
        for (int s = 0; s < warmup_steps; ++s)
            step_fn(sys, a, dt);
    }

    // Timed pass with fresh initial conditions
    System sys = make_solar_system();
    std::vector<Vec3> a = compute_acceleration(sys);
    const double E0 = total_energy(sys);

    auto t0 = std::chrono::high_resolution_clock::now();
    for (int s = 0; s < steps; ++s)
        step_fn(sys, a, dt);
    auto t1 = std::chrono::high_resolution_clock::now();

    double wall = std::chrono::duration<double>(t1 - t0).count();
    double Ef = total_energy(sys);
    double err = std::abs((Ef - E0) / E0);

    std::printf("  [%s] done: %.1f steps/s, |dE/E0|=%.3e, mean=%.4f ms/step\n",
                name.c_str(), steps / wall, err, wall / steps * 1000.0);

    return {name, steps / wall, err, wall, force_evals};
}

int main() {
    constexpr int STEPS        = 1000;
    constexpr int WARMUP_STEPS = 10;
    constexpr double DT        = 5.0;

    std::printf("\n");
    std::printf("============================================================\n");
    std::printf("C++ N-BODY BENCHMARK\n");
    std::printf("Initial condition : solar_system_plus  (%d steps, dt=%.1f days)\n",
                STEPS, DT);
    std::printf("============================================================\n");

    std::vector<BenchResult> results;
    results.push_back(run_bench("Velocity-Verlet", 2, STEPS, WARMUP_STEPS, DT, velocity_verlet));
    results.push_back(run_bench("Ruth-Forest",     3, STEPS, WARMUP_STEPS, DT, ruth_forest));
    results.push_back(run_bench("Yoshida-4",       4, STEPS, WARMUP_STEPS, DT, yoshida4));

    std::printf("\n");
    std::printf("======================================================================\n");
    std::printf("%-20s  %12s  %14s  %18s\n",
                "Integrator", "Steps/sec", "|dE/E0|", "Force evals/step");
    std::printf("----------------------------------------------------------------------\n");
    for (auto& r : results) {
        std::printf("%-20s  %12.1f  %14.3e  %18d\n",
                    r.name.c_str(), r.steps_per_sec, r.energy_error, r.force_evals);
    }
    std::printf("======================================================================\n\n");

    return 0;
}
